src/state/CommitmentAccumulator.sol

```diff
@@ -77,27 +89,39 @@ contract CommitmentAccumulator is ICommitmentAccumulator {
         emit RootAdded(root);
     }
 
-    function _verifyMerkleProof(bytes32 root, bytes32 commitment, bytes32[] calldata path) internal view {
+    function _verifyMerkleProof(
+        bytes32 root,
+        bytes32 commitment,
+        bytes32[] calldata path,
+        uint256 directionBits
+    )
+        internal
+        view
+    {
         // Check length.
-        uint256 pathLength = path.length;
-        uint256 treeDepth = _merkleTree.depth();
-        if (pathLength != treeDepth) {
-            revert InvalidPathLength({ expected: treeDepth, actual: pathLength });
+        if (path.length != _merkleTree.depth()) {
+            revert InvalidPathLength({ expected: _merkleTree.depth(), actual: path.length });
         }
 
         // Check root existence.
         if (!_roots.contains(root)) revert NonExistingRoot(root);
 
-        // Check that the leaf and path reproduce the root.
-        bytes32 computedRoot = path.processProof(commitment, SHA256.commutativeHash);
+        // Check that the commitment leaf and path reproduce the root.
+        bytes32 computedRoot = path.processProof(directionBits, commitment);
+
         if (root != computedRoot) {
             revert InvalidRoot({ expected: root, actual: computedRoot });
         }
     }
 
-    function _merkleProof(bytes32 commitment) internal view returns (bytes32[] memory proof, bytes32 root) {
+    function _merkleProof(bytes32 commitment)
+        internal
+        view
+        returns (bytes32[] memory proof, uint256 directionBits, bytes32 root)
+    {
         uint256 leafIndex = _findCommitmentIndex(commitment);
-        (proof, root) = (_merkleTree.merkleProof(leafIndex), _latestRoot());
+        (proof, directionBits) = (_merkleTree.merkleProof(leafIndex));
+        root = _latestRoot();
```

Comment:
Is the `_latestRoot` the exact current merkle root we are generating the path for? I just want to double-check.

Reply:
Refactored it and added a test in 499aa8ca9bdd803f1559188d8210759bc5927d23